File: apps/web/public/webapp/lib/keyfile.js
// `.flagshipkey` — passphrase-wrapped, portable backup of the User Master Key.
//
// This is the webapp's PRIMARY backup / recovery / cross-device path: the
// browser has no Keychain or iCloud to lean on, so a downloadable encrypted
// key-file is the iCloud stand-in. The 32-byte UMK seed IS the whole account
// (IRK/BAK/SWK/STK all HKDF-derive from it — keystore.js), so anyone holding
// the file AND the passphrase can fully take over the account. Surfaces MUST
// wrap export in heavy warnings and never auto-sync this file anywhere.
//
// Byte-compatible with packages/protocol/src/keyfile.ts and the iOS keyfile
// implementation. The format, AAD canonical string, argon2id KDF, and
// AES-256-GCM AEAD must stay identical so a file written by any surface
// unwraps on every other surface. There is an interop test
// (apps/web/tests/keyfile.test.ts) that asserts a golden file decrypts here.
//
// Crypto:
//   - KDF: argon2id (vendored @noble/hashes — WebCrypto has no argon2).
//     input = UTF8(passphrase); salt = file kdf.saltHex; m/t/p from file;
//     dkLen = 32.
//   - AEAD: AES-256-GCM via crypto.subtle. WebCrypto appends the 16-byte tag
//     to the ciphertext, which is exactly what ciphertextHex carries.
//   - AAD: a canonical UTF8 header string (see aadBytes) bound into the GCM
//     additionalData, so tampering any header field fails decryption.

import { argon2id } from "../vendor/noble-hashes/argon2.js";
import { bytesToHex, hexToBytes } from "../keystore.js";

export const KEYFILE_MAGIC = "flagship-key";
export const KEYFILE_VERSION = 1;

/** Strong interactive default. Recorded in-file so a future version can raise
 *  it and old files still unwrap with their own recorded params. */
export const KEYFILE_ARGON_PARAMS = { m: 65536, t: 3, p: 4 };

/** Floor only — the UI enforces real passphrase strength. */
export const KEYFILE_MIN_PASSPHRASE = 8;

export class KeyfileError extends Error {
  /** @param {string} message @param {"malformed"|"bad-passphrase"|"version"} code */
  constructor(message, code) {
    super(message);
    this.name = "KeyfileError";
    this.code = code;
  }
}

/** Canonical AAD binding the human-meaningful header to the ciphertext. Must
 *  match packages/protocol/src/keyfile.ts aadBytes() byte-for-byte. */
function aadBytes(env) {
  const k = env.kdf;
  const s = [
    "flagship/keyfile/v1",
    String(env.version),
    env.username,
    env.accountId ?? "",
    env.createdAt,
    `${k.algo}|m=${k.m}|t=${k.t}|p=${k.p}`,
    env.aead,
  ].join("|");
  return new TextEncoder().encode(s);
}

function deriveKey(passphrase, salt, p) {
  return argon2id(new TextEncoder().encode(passphrase), salt, {
    m: p.m,
    t: p.t,
    p: p.p,
    dkLen: 32,
  });
}

async function aesGcmEncrypt(keyBytes, nonce, aad, plaintext) {
  const key = await crypto.subtle.importKey("raw", keyBytes, "AES-GCM", false, ["encrypt"]);
  return new Uint8Array(
    await crypto.subtle.encrypt({ name: "AES-GCM", iv: nonce, additionalData: aad }, key, plaintext),
  );
}

async function aesGcmDecrypt(keyBytes, nonce, aad, ciphertext) {
  const key = await crypto.subtle.importKey("raw", keyBytes, "AES-GCM", false, ["decrypt"]);
  return new Uint8Array(
    await crypto.subtle.decrypt({ name: "AES-GCM", iv: nonce, additionalData: aad }, key, ciphertext),
  );
}

function randomBytes(n) {
  const b = new Uint8Array(n);
  crypto.getRandomValues(b);
  return b;
}

/**
 * Wrap a 32-byte UMK seed into `.flagshipkey` text.
 *
 * @param {Uint8Array} seed                 the 32-byte UMK seed
 * @param {string} passphrase               wrap passphrase (>= KEYFILE_MIN_PASSPHRASE)
 * @param {{username:string, accountId?:string, createdAt?:string}} meta
 * @param {{m:number,t:number,p:number}} [argonParams]  injectable for tests
 * @returns {Promise<string>}               the JSON file text (trailing \n)
 */
export async function wrapUmkToKeyfile(seed, passphrase, meta, argonParams = KEYFILE_ARGON_PARAMS) {
  if (!(seed instanceof Uint8Array) || seed.length !== 32) {
    throw new KeyfileError("UMK seed must be 32 bytes", "malformed");
  }
  if (typeof passphrase !== "string" || passphrase.length < KEYFILE_MIN_PASSPHRASE) {
    throw new KeyfileError(`passphrase too short (min ${KEYFILE_MIN_PASSPHRASE})`, "malformed");
  }
  if (!meta || typeof meta.username !== "string" || !meta.username) {
    throw new KeyfileError("username required", "malformed");
  }
  const salt = randomBytes(16);
  const nonce = randomBytes(12);
  const createdAt = meta.createdAt ?? new Date().toISOString();
  const kdf = {
    algo: "argon2id",
    m: argonParams.m,
    t: argonParams.t,
    p: argonParams.p,
    saltHex: bytesToHex(salt),
  };
  const header = {
    version: KEYFILE_VERSION,
    username: meta.username,
    ...(meta.accountId !== undefined ? { accountId: meta.accountId } : {}),
    createdAt,
    kdf,
    aead: "aes-256-gcm",
  };
  const key = deriveKey(passphrase, salt, argonParams);
  const ct = await aesGcmEncrypt(key, nonce, aadBytes(header), seed);
  const env = {
    magic: KEYFILE_MAGIC,
    ...header,
    nonceHex: bytesToHex(nonce),
    ciphertextHex: bytesToHex(ct),
  };
  return `${JSON.stringify(env, null, 2)}\n`;
}

/**
 * Parse + decrypt a `.flagshipkey` file. Throws KeyfileError on any failure:
 *   - "malformed" : not a flagship key file / bad shape
 *   - "version"   : unsupported file version
 *   - "bad-passphrase" : wrong passphrase or tampered/corrupted file
 *
 * @param {string} fileText
 * @param {string} passphrase
 * @returns {Promise<{ seed: Uint8Array, meta: {username:string, accountId?:string, createdAt:string} }>}
 */
export async function unwrapUmkFromKeyfile(fileText, passphrase) {
  let env;
  try {
    env = JSON.parse(fileText);
  } catch {
    throw new KeyfileError("not valid JSON", "malformed");
  }
  if (!env || typeof env !== "object" || env.magic !== KEYFILE_MAGIC) {
    throw new KeyfileError("not a flagship key file", "malformed");
  }
  if (env.version !== KEYFILE_VERSION) {
    throw new KeyfileError(`unsupported version ${env.version}`, "version");
  }
  if (!env.kdf || env.kdf.algo !== "argon2id" || env.aead !== "aes-256-gcm") {
    throw new KeyfileError("unsupported kdf/aead", "malformed");
  }
  if (
    typeof env.kdf.saltHex !== "string" ||
    typeof env.nonceHex !== "string" ||
    typeof env.ciphertextHex !== "string" ||
    typeof env.username !== "string" ||
    typeof env.createdAt !== "string"
  ) {
    throw new KeyfileError("missing required fields", "malformed");
  }
  const salt = hexToBytes(env.kdf.saltHex);
  const nonce = hexToBytes(env.nonceHex);
  const ct = hexToBytes(env.ciphertextHex);
  const key = deriveKey(passphrase, salt, env.kdf);
  let seed;
  try {
    seed = await aesGcmDecrypt(key, nonce, aadBytes(env), ct);
  } catch {
    throw new KeyfileError("wrong passphrase or corrupted/tampered file", "bad-passphrase");
  }
  if (seed.length !== 32) {
    throw new KeyfileError("decrypted seed is not 32 bytes", "malformed");
  }
  return {
    seed,
    meta: {
      username: env.username,
      ...(env.accountId !== undefined ? { accountId: env.accountId } : {}),
      createdAt: env.createdAt,
    },
  };
}

File: apps/mobile/ios/Sources/Flagship/Keyfile.swift
import Foundation
import CryptoKit
import Argon2Kit
import FlagshipCore

/// `.flagshipkey` — a passphrase-wrapped, portable backup of the User
/// Master Key. Byte-compatible with `packages/protocol/src/keyfile.ts`.
///
/// The UMK seed (32 bytes) is the ENTIRE account: IRK/BAK/SWK/STK all
/// HKDF-derive from it, so this file is the keys to the kingdom — anyone
/// with the file AND the passphrase can fully control the account. It is
/// the cloud-independent recovery + cross-device backup path. Surfaces
/// MUST wrap export in heavy warnings and never auto-sync this file.
///
/// Format: JSON, binary fields hex. A self-describing header is bound
/// into the AES-256-GCM AAD, so tampering any header field (username,
/// version, kdf params) fails decryption. The argon2id-derived key
/// (params recorded in-file so they can be raised later without breaking
/// old files) wraps the 32-byte seed.
public enum Keyfile {

    public static let magic = "flagship-key"
    public static let version = 1

    /// memory in KiB (m), iterations (t), parallelism (p).
    public struct ArgonParams: Sendable, Equatable {
        public let m: Int
        public let t: Int
        public let p: Int
        public init(m: Int, t: Int, p: Int) {
            self.m = m; self.t = t; self.p = p
        }
    }

    /// Strong interactive default. Recorded in the file so a future
    /// version can raise it and old files still unwrap with their own
    /// recorded params. Matches `KEYFILE_ARGON_PARAMS` on the TS side.
    public static let argonParams = ArgonParams(m: 65536, t: 3, p: 4)

    /// Floor only — surfaces enforce real passphrase strength in the UI.
    public static let minPassphrase = 8

    public struct Meta: Sendable, Equatable {
        public let username: String
        public let accountId: String?
        /// ISO-8601
        public let createdAt: String
        public init(username: String, accountId: String?, createdAt: String) {
            self.username = username
            self.accountId = accountId
            self.createdAt = createdAt
        }
    }

    public enum KeyfileError: Error, LocalizedError, Equatable {
        /// Not a flagship key file / not valid JSON / structurally bad.
        case malformed(String)
        /// Wrong passphrase or corrupted/tampered ciphertext.
        case badPassphrase
        /// Unsupported version.
        case version(Int)

        public var errorDescription: String? {
            switch self {
            case .malformed(let m): return "This isn't a Flagship key file. (\(m))"
            case .badPassphrase:    return "That passphrase didn't open the file."
            case .version(let v):   return "Unsupported key file version \(v)."
            }
        }
    }

    /// Canonical AAD binding the human-meaningful header to the
    /// ciphertext. MUST match `aadBytes` in keyfile.ts byte-for-byte.
    static func aad(
        version: Int,
        username: String,
        accountId: String?,
        createdAt: String,
        params: ArgonParams
    ) -> Data {
        let s = [
            "flagship/keyfile/v1",
            String(version),
            username,
            accountId ?? "",
            createdAt,
            "argon2id|m=\(params.m)|t=\(params.t)|p=\(params.p)",
            "aes-256-gcm",
        ].joined(separator: "|")
        return Data(s.utf8)
    }

    /// argon2id KDF. input = UTF8(passphrase); dkLen = 32. Maps the
    /// file's m (memory KiB) / t (iterations) / p (parallelism) onto the
    /// reference C library via Argon2Kit. Pinned to argon2id + V13.
    static func deriveKey(passphrase: String, salt: Data, params: ArgonParams) throws -> SymmetricKey {
        let digest = try Argon2.hash(
            password: Data(passphrase.utf8),
            salt: salt,
            iterations: UInt32(params.t),
            memory: UInt32(params.m),
            threads: UInt32(params.p),
            length: 32,
            type: .id,
            version: .v13
        )
        return SymmetricKey(data: digest.rawData)
    }

    // MARK: - Wrap (export)

    /// Wrap a 32-byte UMK seed into `.flagshipkey` text. `params` is
    /// injectable for tests; production callers should omit it.
    public static func wrap(
        umkSeed: Data,
        passphrase: String,
        meta: Meta,
        params: ArgonParams = Keyfile.argonParams
    ) throws -> String {
        guard umkSeed.count == 32 else {
            throw KeyfileError.malformed("UMK seed must be 32 bytes")
        }
        guard passphrase.count >= minPassphrase else {
            throw KeyfileError.malformed("passphrase too short (min \(minPassphrase))")
        }
        var saltBytes = [UInt8](repeating: 0, count: 16)
        _ = saltBytes.withUnsafeMutableBytes { SecRandomCopyBytes(kSecRandomDefault, 16, $0.baseAddress!) }
        let salt = Data(saltBytes)

        let nonceBytes = (0..<12).map { _ in UInt8.random(in: 0...255) }
        let nonceData = Data(nonceBytes)
        let nonce = try AES.GCM.Nonce(data: nonceData)

        let key = try deriveKey(passphrase: passphrase, salt: salt, params: params)
        let aadData = aad(
            version: version,
            username: meta.username,
            accountId: meta.accountId,
            createdAt: meta.createdAt,
            params: params
        )
        let sealed = try AES.GCM.seal(umkSeed, using: key, nonce: nonce, authenticating: aadData)
        // TS ciphertextHex = plaintext-ciphertext + 16-byte GCM tag.
        let ciphertext = sealed.ciphertext + sealed.tag

        return envelopeJSON(
            meta: meta,
            params: params,
            saltHex: HexUtil.encode(salt),
            nonceHex: HexUtil.encode(nonceData),
            ciphertextHex: HexUtil.encode(ciphertext)
        )
    }

    // MARK: - Unwrap (import)

    /// Wrap a UMK seed held as a CryptoKit SymmetricKey.
    public static func wrap(
        umkSeed: SymmetricKey,
        passphrase: String,
        meta: Meta,
        params: ArgonParams = Keyfile.argonParams
    ) throws -> String {
        try wrap(
            umkSeed: umkSeed.withUnsafeBytes { Data($0) },
            passphrase: passphrase,
            meta: meta,
            params: params
        )
    }

    /// Parse + decrypt a `.flagshipkey` file. Throws KeyfileError on
    /// any failure (malformed / wrong-passphrase / version).
    public static func unwrap(
        fileText: String,
        passphrase: String
    ) throws -> (umkSeed: SymmetricKey, meta: Meta) {
        guard let data = fileText.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw KeyfileError.malformed("not valid JSON")
        }
        guard (obj["magic"] as? String) == magic else {
            throw KeyfileError.malformed("not a flagship key file")
        }
        guard let ver = obj["version"] as? Int else {
            throw KeyfileError.malformed("missing version")
        }
        guard ver == version else {
            throw KeyfileError.version(ver)
        }
        guard let kdf = obj["kdf"] as? [String: Any],
              (kdf["algo"] as? String) == "argon2id",
              (obj["aead"] as? String) == "aes-256-gcm" else {
            throw KeyfileError.malformed("unsupported kdf/aead")
        }
        guard let m = intValue(kdf["m"]),
              let t = intValue(kdf["t"]),
              let p = intValue(kdf["p"]),
              let saltHex = kdf["saltHex"] as? String,
              let username = obj["username"] as? String,
              let createdAt = obj["createdAt"] as? String,
              let nonceHex = obj["nonceHex"] as? String,
              let ciphertextHex = obj["ciphertextHex"] as? String,
              let salt = HexUtil.decode(saltHex),
              let nonceData = HexUtil.decode(nonceHex),
              let ctAndTag = HexUtil.decode(ciphertextHex) else {
            throw KeyfileError.malformed("missing or malformed fields")
        }
        guard ctAndTag.count >= 16 else {
            throw KeyfileError.malformed("ciphertext too short")
        }
        let accountId = obj["accountId"] as? String
        let params = ArgonParams(m: m, t: t, p: p)

        let key = try deriveKey(passphrase: passphrase, salt: salt, params: params)
        let aadData = aad(
            version: ver,
            username: username,
            accountId: accountId,
            createdAt: createdAt,
            params: params
        )
        // ciphertextHex is plaintext-ciphertext + 16B tag appended.
        let tag = ctAndTag.suffix(16)
        let ciphertext = ctAndTag.prefix(ctAndTag.count - 16)
        do {
            let nonce = try AES.GCM.Nonce(data: nonceData)
            let box = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag)
            let seed = try AES.GCM.open(box, using: key, authenticating: aadData)
            guard seed.count == 32 else {
                throw KeyfileError.malformed("decrypted seed is not 32 bytes")
            }
            return (
                SymmetricKey(data: seed),
                Meta(username: username, accountId: accountId, createdAt: createdAt)
            )
        } catch let e as KeyfileError {
            throw e
        } catch {
            throw KeyfileError.badPassphrase
        }
    }

    // MARK: - Helpers

    /// Build the envelope JSON. We emit it by hand (rather than
    /// JSONEncoder) to guarantee the field set + accountId-omission
    /// match the TS writer's shape exactly.
    private static func envelopeJSON(
        meta: Meta,
        params: ArgonParams,
        saltHex: String,
        nonceHex: String,
        ciphertextHex: String
    ) -> String {
        var root: [String: Any] = [
            "magic": magic,
            "version": version,
            "username": meta.username,
            "createdAt": meta.createdAt,
            "kdf": [
                "algo": "argon2id",
                "m": params.m,
                "t": params.t,
                "p": params.p,
                "saltHex": saltHex,
            ],
            "aead": "aes-256-gcm",
            "nonceHex": nonceHex,
            "ciphertextHex": ciphertextHex,
        ]
        if let accountId = meta.accountId {
            root["accountId"] = accountId
        }
        let data = (try? JSONSerialization.data(
            withJSONObject: root,
            options: [.prettyPrinted, .sortedKeys]
        )) ?? Data()
        let json = String(data: data, encoding: .utf8) ?? "{}"
        return json + "\n"
    }

    private static func intValue(_ any: Any?) -> Int? {
        if let i = any as? Int { return i }
        if let n = any as? NSNumber { return n.intValue }
        if let d = any as? Double { return Int(d) }
        return nil
    }
}
